Smoke tests

sgpearse opened the pull request:
This PR consists of two smoke tests, 1) testGrid, and 2) testDataMgr
 
Like all of our test applications, compiling these tools need **BUILD_TEST_APPS** to be switched **ON** with ccmake.

See the headings of testGrid.cpp and testDataMgr.cpp for what the tests are exercising.

Both tools have flags that can be listed with the -help option.  Issuing -help for testGrid yields the following:

```
cisl-farmington:build pearse$ bin/testGrid -help
    OPTION          NUM_ARGS    DEFAULT
    ------          --------    -------
    -arrangements   1           Constant:Ramp:Triangle
        Colon delimited list of data arrangements to test synthetic grids with
    -bs             1           64:64:64
        Internal storage blocking factor expressed in grid points (NX:NY:NZ)
    -dims           1           8:8:8
        Data volume dimensions expressed in grid points (NX:NY:NZ)
    -extents        1           0:0:0:1:1:1
        Colon delimited 6-element vector specifying domain extents of synthetic
        grids in user coordinates (X0:Y0:Z0:X1:Y1:Z1)
    -grids          1           Regular:Stretched:Layered:Curvilinear
        Colon delimited list of grids to test
    -help           0           false
        Print this message and exit
    -maxValue       1           1
        The maximum data value to be assigned to cells in synthetic grids
    -minValue       1           0
        The minimum data value to be assigned to cells in synthetic grids

Copyright 2007 The National Center for Atmospheric Research

Version: 3.2.0 (2020-03-09T21:46:32Z) www.vapor.ucar.edu
```

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,439 @@
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+ 
+    // Layered and Curvilinear grids have a bug where querying indices 
```
What is the bug? How do you reproduce it?

sgpearse replied:
The bug is that Grid::GetValueAtIndex (and AccessIJK) reports wrong indices for Layered and Curvilinear grids.

I've made some changes to make testing this more clear.  After pulling and building the current code base, run the following commands

testGrid -arrangement Ramp

This will test coordinate vs IJK GetValue() calls for:
1) linear interpolation
2) nearest neighbor
3) nearest neighbor with nudging

Take a look at the results for the layered grid, which I've pasted below.  Nudging fixes the RMS error in this case.


Layered 8x8x8 Ramp up on Z axis:
  linear:
    RMS error:                                     0.1830
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  45

  nearestNeighbor:
    RMS error:                                     0.1603
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  90

  nearestNeighbor (nudged):
    RMS error:                                     0.0000
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  0

sgpearse commented:
@clyne Sorry John, I was not clear on how to use the driver.

Now that issue #2252 is fixed, I've added baseline files to this PR that we can use to compare test results with.  If the Grid or DataMgr tests differ from the baseline, the test will exit -1.

I've also added logic to automatically generate baseline files if none are detected.

You will need to download the test data linked here: https://drive.google.com/open?id=1w8CLOohQuVrhcDbmIyU68whvqaoiCx9t

smokeTests.py will also need you to pass two arguments: --testDataRoot and --binaryRoot.  Passing the --help flag yields the following:

`cisl-farmington:VAPOR_gridTests pearse$ ./test_apps/smokeTests/smokeTests.py --help
usage: A test driver for the DataMgr and Grid classes [-h]
                                                      [--makeBaseline false]
                                                      [--testDataRoot /path/to/data]
                                                      [--binaryRoot /path/to/binaries]

optional arguments:
  -h, --help            show this help message and exit
  --makeBaseline false  Boolean that makes these test results the baseline on
                        which future tests will be compared. If no baseline
                        file exists, this will automatically be set to true.
  --testDataRoot /path/to/data
                        Directory where DataMgr test data is stored.
  --binaryRoot /path/to/binaries
                        Directory where binary test programs (testGrid,
                        testDataMgr) are stored.`

sgpearse commented:
@clyne After rebuilding the baseline files, I believe another bug has arisen.  See #2259.

We shouldn't accept this PR until we have good baseline files to compare future changes with.

sgpearse commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,358 @@
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                value = value==minVal ? maxVal : minVal;
```
Yes, a triangle signal.  Up, down, up, down etc...

sgpearse commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,358 @@
+    }
+}
+
+void MakeRamp( Grid* grid, float minVal, float maxVal, size_t axis=X ) {
```
Got it.  I'll add this to my to-do list.

sgpearse replied:
Done.

sgpearse commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,358 @@
+    }
+}
+
+void CompareIndexToCoords( 
```
Done.

sgpearse commented on test_apps/smokeTests/testGrid.cpp:
```diff
@@ -0,0 +1,248 @@
+    cout << "Elapsed time: " << t1-t0 << endl;
+
+    DeleteHeap();
+}
```
Done.

clyne commented on test_apps/smokeTests/smokeTests.py:
```diff
@@ -0,0 +1,237 @@
+#! /usr/local/bin/python3
+#! /usr/bin/python3
```
Can you have two shebangs in a script? Why not simply use the 'env' command for portability:

#!/usr/bin/env python3 

should work just about everywhere

sgpearse replied:
Thanks - Did not know that.

clyne commented on test_apps/smokeTests/testDataMgr.cpp:
```diff
@@ -0,0 +1,136 @@
+        cout << fileName << endl;
+    }
+
+    TestDataMgr( opt.fileType, opt.memsize, opt.nthreads, files, options );
```
Check return status of TestDataMgr

sgpearse replied:
Done.

clyne commented on test_apps/smokeTests/testDataMgr.cpp:
```diff
@@ -0,0 +1,136 @@
+    TestDataMgr( opt.fileType, opt.memsize, opt.nthreads, files, options );
+
+    cout << "Elapsed time: " << Wasp::GetTime() - t0 << endl;
+}
```
No return status from main

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/testGrid.cpp:
```diff
@@ -0,0 +1,274 @@
+
+    DeleteHeap();
+
+    return rc;
```
The convention for executables is to return 0 for success and 1 for failure. For whatever reasons system calls return a negative int on failure, but executables return a value between 0 and 255, with 0 indicating success and 1 indicating failure. The remaining 254 values are reserved for other forms of status.

sgpearse replied:
Thanks - fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,436 @@
+
+// This function iterates across all nodes in a grid; making comparisons between the
+// data values returned by the functions GetValueAtIndex() and GetValue().
+int CompareIndexToCoords( 
```
The convention we use for error handling throughout VAPOR is that functions that report errors via MyBase return integer status (or NULL in the case of pointers), and functions that do not report errors return a boolean to indicate success or failure.

sgpearse replied:
Fixed

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,436 @@
+
+// This function iterates across all nodes in a grid; making comparisons between the
+// data values returned by the functions GetValueAtIndex() and GetValue().
+int CompareIndexToCoords( 
```
How about using a consistent naming convention for the unit tests. You have TestNodeIterator below, so why not something like TestIndexCoordConsistency?

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,436 @@
+    return 0;
+}
+
+size_t TestNodeIterator( const Grid *g, int& count, double& time ) {
```
Rather than returning a value that has to be interpreted to determine success or failure, how about simply adopting the convention that all TestXXX functions return a boolean indicating success or failure?

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,436 @@
+
+    return expectedCount;
+}
+
```
It would be good to have some additional iterator tests. We have three iterator types:

Grid::Iterator // dereferences data values
Grid::ConstCoordItr // dereferences node coordinates
Grid::ConstNodeIterator // dereferences node indices

A simple test for each would be to verify that the dereferenced values are the same as what you get with explicit access using the accessors.

sgpearse replied:
@clyne - Done.  I believe there are bugs in the ConstCoordIterator, for the Regular and Layered grids.  Corresponding with more details over email.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,436 @@
+    return rc;
+}
+
+int RunTests( 
```
Follow VAPOR conventions for return status; use booleans if errors aren't logged via MyBase

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/testGrid.cpp:
```diff
@@ -0,0 +1,274 @@
+    if (opt.dims.size() != 3) {
+        cerr << "The -dims flag must contain 3 elements if used" << endl;
+        op.PrintOptionHelp(stderr);
+        exit(0);
```
Exit with a 1, not zero for abnormal termination

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/testGrid.cpp:
```diff
@@ -0,0 +1,274 @@
+
+    if (opt.help) {
+        op.PrintOptionHelp(stderr);
+        exit(0);
```
exit codes

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/testGrid.cpp:
```diff
@@ -0,0 +1,274 @@
+
+    if ( regularRC != 0 ) {
+        cout << "Errors occurred while testing Grid::RegularGrid." << endl;
+        rc = -1;
```
Follow exit code conventions for programs; return 0 on success and 1 on failure

sgpearse replied:
Thanks - I confused return codes with exit codes here.

sgpearse commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,358 @@
+                    disagreements++;
+                }
+
+                if (sampleValue == grid->GetMissingValue()) {
```
Done.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+
+size_t TestConstNodeIterator( 
+    const Grid *g, 
+    size_t& count, 
```
Is there a reasonable use case when count, disagreements, or time  are not going to be initialized to zero by the caller? If not, initialize it inside your test function.

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+    std::vector< size_t > dims = g->GetDimensions();
+
+    for ( ; itr!=enditr; ++itr) {
+        size_t i = count % dims[X];
```
The utility function Wasp::VectorizeCoords will turn an offset into component coordinates and will not break if the grid dimension is not 2 or 3 as this code assumes.

sgpearse replied:
Good to know.  Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+  return std::abs(x - y) > epsilon * std::abs(x);
+}
+
+size_t TestConstNodeIterator( 
```
Per earlier comments having the test function return a boolean to indicate success or failure (rather than an  would make the status of the test much more clear. It's really hard to understand under what conditions this test fails by reading the code; it returns expectedCount, that has nothing to do with what is being tested.

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+    return expectedCount;
+}
+
+size_t TestIterator( 
```
Same comments as for TestConstNodeIterator

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+    return expectedCount;
+}
+
+size_t TestConstCoordItr( 
```
Same comments as for TestConstNodeIterator

sgpearse replied:
Fixed

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+    size_t disagreements;
+
+    if ( CompareIndexToCoords( grid, rms, numMissingValues, disagreements ) ) {
+        cout << "       *** Error reported in " << grid->GetType() << " grid***" << endl;
```
errors to std::cerr

sgpearse replied:
Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+
+    if ( CompareIndexToCoords( grid, rms, numMissingValues, disagreements ) ) {
+        cout << "       *** Error reported in " << grid->GetType() << " grid***" << endl;
+        rc = 1;
```
Let's return true on success, not false :-)

sgpearse replied:
Oops.  Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+       
+        grid->SetInterpolationOrder( linear );
+        if ( RunTest( grid ) != 0 )
+            rc = 1;
```
Return true on success, false on failure per our error conventions and common practice.

sgpearse replied:
Fixed.

sgpearse commented on test_apps/smokeTests/dataMgrTools.cpp:
```diff
@@ -0,0 +1,167 @@
+    VAPoR::DataMgr dataMgr(fileType, memsize, nthreads);
+    int rc = dataMgr.Initialize(files, options);
+    if (rc<0) {
+        cout << "Failed to intialize WRF DataMGR" << endl;
```
Done.

sgpearse commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,594 @@
+    rms = sqrt( sum / (x*y*z) );
+
+    if ( rms != 0 || disagreements > 0 )
+        return 1;
```
Fixed.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,620 @@
+        std::vector<size_t> ijk = Wasp::VectorizeCoords( count, dims );
+       
+        double itrData  = g->GetValueAtIndex( (*itr).data() );
+        double gridData = g->AccessIJK( ijk[X], ijk[Y], ijk[Z] ); 
```
This will result in a memory access violation if ijk does not have at least three elements. Suggest you use GetValueAtIndex instead of AccessIJK. The former you can simply pass ijk directly (it accepts a 'const std::vector <size_t> &' argument). The latter should be deprecated, and simply calls the former.

sgpearse replied:
done

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,620 @@
+    size_t& count, 
+    size_t& expectedCount,
+    size_t& disagreements, 
+    double& time 
```
time still isn't initialized anywhere.

sgpearse replied:
I don't believe time needs to be initialized.  It just gets assigned to at the end of the function.  

It is derived from t0, which is initialized.

However t0's initialization should probably be grouped with the other variable initializations.  I'll move it up.  If I'm missing something here please let me know.

clyne replied:
Sorry, I missed it. However, I think it is good practice to initialize any output parameters to "reasonable values" up front in the function. Doing so makes it clear to the reader that these arguments are being initialized, and if the code gets refactored later there is less of a chance of an early return being introduced that misses the initialization.

sgpearse replied:
Looking back on it, I agree.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,620 @@
+    size_t& count, 
+    size_t& expectedCount,
+    size_t& disagreements,
+    double& time 
```
time still isn't initialized anywhere.

clyne commented on test_apps/smokeTests/gridTools.cpp:
```diff
@@ -0,0 +1,620 @@
+    size_t& count,
+    size_t& expectedCount,
+    size_t& disagreements,
+    double& time 
```
time still isn't initialized anywhere.
